src/summaries.rs

```diff
@@ -0,0 +1,185 @@
+/// argument values and then simplifying the resulting values under the current path condition.
+///
+/// Summaries are stored in a persistent, per project database. When a crate is recompiled,
+/// all summaries arising from the crate are recomputed, but the database is only updated when
+/// a function summary changes. When this happens, all callers of the function need to be reanalyzed
+/// using the new summary, which could result in their summaries being updated, and so on.
+/// In the case of recursive loops, a function summary may need to be recomputed and widened
+/// until a fixed point is reached. Since crate dependencies are acyclic, fixed point computation
+/// can be limited to the functions of one crate and the dependency graph need not be stored
+/// in the database.
+///
+/// There are three ways summaries are constructed:
+/// 1) By analyzing the body of the actual function.
+/// 2) By analyzing the body of a contract function that contains only enough code to generate
+///    an accurate summary. This is the preferred way to deal with abstract and foreign functions
+///    where the actual body is not known until runtime.
+/// 3) By constructing a dummy summary using only the type signature of the function.
+///    In such cases there are no preconditions, no post conditions, the result value is fully
+///    abstract as is the unwind condition and the values assigned to any mutable parameter.
+///    This makes the summary a conservative over approximation of the actual behavior. It is not
+///    sound, however, if there are side effects on static state since it is neither practical nor
+///    desirable to havoc all static variables every time such a function is called. Consequently
+///    sound analysis is only possible one can assume that all such functions have been provided
+///    with explicit contract functions.
+#[derive(Serialize, Deserialize, Debug, Default, Eq, PartialEq)]
+pub struct Summary {
+    // Conditions that should hold prior to the call.
+    // Callers should substitute parameter values with argument values and simplify the results
+    // under the current path condition. Any values that do not simplify to true will require the
+    // caller to either generate an error message or to add a precondition to its own summary that
+    // will be sufficient to ensure that all of the preconditions in this summary are met.
+    pub preconditions: List<AbstractValue>,
+
+    // If the function returns a value, this summarizes what is known statically of the return value.
+    // Callers should substitute parameter values with argument values and simplify the result
+    // under the current path condition.
+    pub result: Option<AbstractValue>,
+
+    // Modifications the function makes to mutable state external to the function.
+    // Every path will be rooted in a static or in a mutable parameter.
+    // No two paths in this collection will lead to the same place in memory.
```

Comment:
No two paths or no two *distinct* paths? And if the former, would it make sense to use a map to enforce the  invariant that keys or disjoint (or wrap `List<(Path, AbstractValue)>` in something that enforces this)?

Reply:
Good point. Let me think about this a bit more.